psydac/core/bsplines_pyccel.py

```diff
@@ -0,0 +1,1018 @@
+
+    periodic : bool
+        True if domain is periodic, False otherwise.
+
+    multiplicity : int
+        Multiplicity of the knots in the knot sequence, we assume that the same
+        multiplicity applies to each interior knot.
+
+    tol: float
+        If the distance between two knots is less than tol, we assume
+        that they are repeated knots which correspond to the same break point.
+
+    out : array
+        The result will be inserted into this array.
+        It should be of the appropriate shape and dtype.
+    """
+
+    if periodic:
+        T, p = knots, degree
+        period = T[len(knots) -1 - p] - T[p]
+        left   = T[len(knots) -2 - 2 * p] - period
+        right  = T[2 * p + 1] + period
+
+        out[0] = left
+        out[-1] = right
+        out[1: - 1] = knots
+
+    else:
+        out[0] = knots[0]
+        out[1:degree + 2] = knots[:degree+1]
+
+        n_out = out.shape[0]
+        n_knots = len(knots)
+        out[n_out - degree - 2] = knots[n_knots - 1]
+        out[n_out - degree - 1:n_out] = knots[n_knots - degree - 1:]
+
+        if len(knots[degree + 1:n_knots - degree - 1]) > 0:
+            out[degree + 2: degree + 2 + multiplicity] = knots[degree + 1]
+
+            unique_index = 0
+
+            for i in range(degree + 1, len(knots) - degree - 2, 1):
+                if knots[i + 1] - knots[i] > tol:
+                    out[degree + 2 + multiplicity * (unique_index + 1):
+                        degree + 2 + multiplicity * (unique_index + 2)] = knots[i + 1]
+                    unique_index += 1
+
+        else:
+            out[degree + 2: n_out - degree - 2] = knots[degree + 1:n_knots - degree - 1]
+
+
+# =============================================================================
+def quadrature_grid_p(breaks: 'float[:]', quad_rule_x: 'float[:]', quad_rule_w: 'float[:]', out: 'float[:,:,:]'):
+    """
+    Compute the quadrature points and weights for performing integrals over
+    each element (interval) of the 1D domain, given a certain Gaussian
+    quadrature rule.
+
+    An n-point Gaussian quadrature rule for the canonical interval :math:`[-1,+1]`
+    and trivial weighting function :math:`\\omega(x)=1` is defined by the n abscissas
+    :math:`x_i` and n weights :math:`w_i` that satisfy the following identity for
+    polynomial functions :math:`f(x)` of degree :math:`2n-1` or less:
+
+    .. math :: \\int_{-1}^{+1} f(x) dx = \\sum_{i=0}^{n-1} w_i f(x_i)
+
+    Parameters
+    ----------
+    breaks : array_like of floats
+        Coordinates of spline breakpoints.
+
+    quad_rule_x : array_like of ints
+        Coordinates of quadrature points on canonical interval [-1,1].
+
+    quad_rule_w : array_like of ints
+        Weights assigned to quadrature points on canonical interval [-1,1].
+
+    out : array
+        The result will be inserted into this array.
+        It should be of the appropriate shape and dtype.
```

Comment:
Why do you store the quadrature points and weights in the same 3D array, instead of using two separate 2D arrays as in the original function?

Reply:
Changed the docstring accordingly in 9e1ad7c

Reply:
Some more minor changes in 5a49830